Add configurable auto failure handling capability

veerajsk opened the pull request:
Add a AutoFailureHandling config that can be used to enable/disable
auto failure handling for a plan in the engine

veerajsk committed:
Add configurable auto failure handling capability

Add a AutoFailureHandling config that can be used to enable/disable
auto failure handling for a plan in the engine

veerajsk committed:
Add config to all alica.yaml files

abhi1287 commented on alica_engine/src/engine/RuleBook.cpp:
```diff
@@ -228,6 +229,8 @@ PlanChange RuleBook::authorityOverrideRule(RunningPlan& r)
 {
+    if (!_autoFailureHandlingEnabled)
+        return PlanChange::NoChange;
     assert(!r.isRetired());
```
This is tricky, shouldn't we abort the plan execution if the assignment is no longer valid or the failure state is reached? It doesn't fall under auto recovery actions afaik.

veerajsk replied:
my bad, I didn't want to disable the planAbortRule but forgot to remove this piece of code. The rule just marks the running plan as failure but does not handle it in any way. The only thing it seems to do is return PlanChange::Failure, so that the higher level plan can react to the failure in the same tick() of the running plan, so I think its ok to not disable the rule.

veerajsk replied:
I did enable this rule & the tests still work as expected

abhi1287 commented on alica_tests/src/test/test_failure_handling.cpp:
```diff
@@ -43,5 +81,110 @@ TEST_F(AlicaFailureHandlingEnabledFixture, redoPlanOnFailure)
+    STEP_UNTIL(test::Util::isStateActive(ae, FAILURE_PLAN_FAILED_STATE));
+    ASSERT_TRUE(test::Util::isStateActive(ae, FAILURE_PLAN_FAILED_STATE));
+
+    // Check if we remain in the failed state
```
unset the transition conditions so that plan redo doesn't take the agent back to failure state.

veerajsk replied:
done

abhi1287 commented on alica_tests/src/test/test_failure_handling.cpp:
```diff
@@ -43,5 +81,110 @@ TEST_F(AlicaFailureHandlingEnabledFixture, redoPlanOnFailure)
+    wm0->setTransitionCondition1023566846009251524(true);
+    wm0->setTransitionCondition1446293122737278544(false);
+    wm1->setTransitionCondition1446293122737278544(false);
+    STEP_UNTIL(acs[0], test::Util::isStateActive(aes[0], FAILURE_PLAN_INIT_STATE));
```
ideally plan init should increment a static counter so that you can be sure that plan wasn't restarted and only ran once. otherwise step_until is just gonna step until the plan potentially restarts and fail again.

veerajsk replied:
done

veerajsk committed:
Enable plan abort & make tests more robust

The plan abort rule should not be considered an auto failure handling
behaviour so that the failure can be immediately handled by the higher
levels.

Some of the tests could pass if a plan restarted when they should fail,
this is suitably addressed by ensuring transitions are unset when they
are no longer needed to be enabled & by ensuring the plan init is only
executed once
